<!DOCTYPE html>
<html>

<head>
    <meta charset='utf-8'>
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <meta name="description" content="Biojs-vis-proteinfeaturesviewer : A BioJS viewer for protein sequence features">

    <link rel="stylesheet" type="text/css" media="screen" href="stylesheets/stylesheet.css">

    <title>Developer Guide</title>
</head>

<body>

<!-- HEADER -->
<div id="header_wrap" class="outer">
    <header class="inner">
        <a href="."><h1 id="project_title">UniProt Feature Viewer</h1></a>
        <h2 id="project_tagline">A BioJS viewer for protein sequence features</h2>

        <section id="links">
            <a class="up_links" id="forkme_banner"
               href="https://github.com/ebi-uniprot/biojs-vis-proteinFeaturesViewer">View on GitHub</a>
            <a class="up_links" href="./developerGuide.html">Developer Guide</a>
            <a class="up_links" href="./userGuide.html">User Guide</a>
        </section>
    </header>
</div>

<!-- MAIN CONTENT -->
<div id="main_content_wrap" class="outer">
    <section id="main_content" class="inner">
<div class="container">
    <div class="static-content">
        <h1 id="biojs-vis-proteinfeaturesviewer">biojs-vis-proteinFeaturesViewer</h1>

<h2 id="getting-the-code">Getting the code</h2>
<p>As well as <code class="highlighter-rouge">featuresviewer.min.js</code> which contains the minified JavaScript code required to run the application, two css files are required: <code class="highlighter-rouge">main.css</code> (general styling, including feature styles) and <code class="highlighter-rouge">fontello.css</code> (icon font).</p>

<h3 id="the-easy-way-cdn">The easy way (CDN)</h3>
<p>We host all our releases on Github, using Github pages as a CDN. The latest release is available here:</p>

<div class="highlighter-rouge"><pre class="highlight"><code>&lt;script src="http://ebi-uniprot.github.io/CDN/feature-viewer/featuresviewer.js"&gt;&lt;/script&gt;
&lt;link href="http://ebi-uniprot.github.io/CDN/feature-viewer/css/main.css" rel="stylesheet"/&gt;
&lt;link href="http://ebi-uniprot.github.io/CDN/feature-viewer/css/fontello.css" rel="stylesheet"/&gt;   
</code></pre>
</div>

<p>We also archive previous versions:</p>

<div class="highlighter-rouge"><pre class="highlight"><code>&lt;script src="http://ebi-uniprot.github.io/CDN/feature-viewer/1.0.0/featuresviewer.js"&gt;&lt;/script&gt;
&lt;link href="http://ebi-uniprot.github.io/CDN/feature-viewer/1.0.0/css/main.css" rel="stylesheet"/&gt;
&lt;link href="http://ebi-uniprot.github.io/CDN/feature-viewer/1.0.0/css/fontello.css" rel="stylesheet"/&gt;   
</code></pre>
</div>

<h3 id="the-node-way">The node way</h3>
<p><code class="highlighter-rouge">&gt;npm install biojs-vis-proteinfeaturesviewer</code><br />
This is probably the preferred way if you are developing a new component or application that uses our feature viewer. It would be useful as well if you are modifying this component. Some more information about <a href="https://nodejs.org/en/">node</a> and <a href="https://www.npmjs.com/">NPM</a>.</p>

<h2 id="loading-the-feature-viewer">Loading the Feature Viewer</h2>
<p>In order to use this component, you need an HTML element such as &lt;div&gt; or &lt;span&gt; or anything else that can act as a container where the viewer will be located. In the following lines, we use a DIV element identified as <em>yourDiv</em>, it could have any other name.</p>

<p>When creating the instance, you need to specify the object where the component will be located, <em>el</em>, as well as the UniProt accession whose sequence annotations will be visualized, <em>uniprotacc</em>. We are supposing here that you will instantiate the component as soon as the window has been loaded, i.e., window.onload function.</p>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="nt">&lt;div</span> <span class="na">id=</span><span class="s">'yourDiv'</span><span class="nt">/&gt;</span>
<span class="nt">&lt;script&gt;</span>
    <span class="nb">window</span><span class="p">.</span><span class="nx">onload</span> <span class="o">=</span> <span class="kd">function</span><span class="p">()</span> <span class="p">{</span>
        <span class="kd">var</span> <span class="nx">yourDiv</span> <span class="o">=</span> <span class="nb">document</span><span class="p">.</span><span class="nx">getElementById</span><span class="p">(</span><span class="s1">'yourDiv'</span><span class="p">);</span>
        <span class="kd">var</span> <span class="nx">biojs_vis_proteinFeaturesViewer</span> <span class="o">=</span> <span class="nx">require</span><span class="p">(</span><span class="s1">'biojs-vis-proteinfeaturesviewer'</span><span class="p">);</span>
        <span class="kd">var</span> <span class="nx">instance</span> <span class="o">=</span> <span class="k">new</span> <span class="nx">biojs_vis_proteinFeaturesViewer</span><span class="p">({</span>
            <span class="na">el</span><span class="p">:</span> <span class="nx">yourDiv</span><span class="p">,</span>
            <span class="na">uniprotacc</span> <span class="p">:</span> <span class="s1">'P05067'</span>
        <span class="p">});</span>
    <span class="p">}</span>
<span class="nt">&lt;/script&gt;</span>
</code></pre>
</div>

<p>That’s it, you should now see the Feature Viewer in your web page!</p>

<h2 id="api">API</h2>
<p>The protein feature viewer offer some public methods that can be used to programmatically interact with the component.</p>

<h3 id="selectfeaturefttype-begin-end-altsequence">selectFeature(ftType, begin, end, altSequence)</h3>
<p>If you want to select a feature you can use the <em>selectFeature method</em>. The parameters ftType, begin and end are mandatory while altSequence is optional. Whenever you want to select variants, mutagenesis or conflict sequences you need to pass the altSequence, i.e., the bit of sequence reported by that feature.</p>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="o">&lt;</span><span class="nx">script</span><span class="o">&gt;</span>
    <span class="nx">instance</span><span class="p">.</span><span class="nx">selectFeature</span><span class="p">(</span><span class="s1">'region'</span><span class="p">,</span> <span class="mi">27</span><span class="p">,</span> <span class="mi">83</span><span class="p">);</span>
    <span class="nx">instance</span><span class="p">.</span><span class="nx">selectFeature</span><span class="p">(</span><span class="s1">'act_site'</span><span class="p">,</span> <span class="mi">47</span><span class="p">,</span> <span class="mi">47</span><span class="p">);</span>
<span class="o">&lt;</span><span class="sr">/script</span><span class="err">&gt;
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="o">&lt;</span><span class="nx">script</span><span class="o">&gt;</span>
    <span class="nx">instance</span><span class="p">.</span><span class="nx">selectFeature</span><span class="p">(</span><span class="s1">'variant'</span><span class="p">,</span> <span class="mi">33</span><span class="p">,</span> <span class="mi">34</span><span class="p">,</span> <span class="s1">'QH'</span><span class="p">);</span>
    <span class="nx">instance</span><span class="p">.</span><span class="nx">selectFeature</span><span class="p">(</span><span class="s1">'conflict'</span><span class="p">,</span> <span class="mi">53</span><span class="p">,</span> <span class="mi">58</span><span class="p">,</span> <span class="s1">'KGLMTW'</span><span class="p">);</span>
    <span class="nx">instance</span><span class="p">.</span><span class="nx">selectFeature</span><span class="p">(</span><span class="s1">'mutagen'</span><span class="p">,</span> <span class="mi">115</span><span class="p">,</span> <span class="mi">115</span><span class="p">,</span> <span class="s1">'K'</span><span class="p">);</span>
<span class="o">&lt;</span><span class="sr">/script</span><span class="err">&gt;
</span></code></pre>
</div>

<p>Here is the list of the currently supported feature types:</p>

<p><strong>Category: Domains and sites</strong></p>

<table>
  <thead>
    <tr>
      <th><strong>ftType</strong></th>
      <th><strong>Label</strong></th>
      <th><strong>Description</strong></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><a href="http://www.uniprot.org/help/domain">domain</a></td>
      <td>Domain</td>
      <td>Position and type of each modular protein domain</td>
    </tr>
    <tr>
      <td><a href="http://www.uniprot.org/help/repeat">repeat</a></td>
      <td>Repeat</td>
      <td>Positions of repeated sequence motifs or repeated domains</td>
    </tr>
    <tr>
      <td><a href="http://www.uniprot.org/help/ca_bind">ca_bind</a></td>
      <td>Calcium binding</td>
      <td>Position(s) of calcium binding region(s) within the protein</td>
    </tr>
    <tr>
      <td><a href="http://www.uniprot.org/help/zn_fing">zn_fing</a></td>
      <td>Zinc finger</td>
      <td>Position(s) and type(s) of zinc fingers within the protein</td>
    </tr>
    <tr>
      <td><a href="http://www.uniprot.org/help/dna_bind">dna_bind</a></td>
      <td>DNA binding</td>
      <td>Position and type of a &lt;span class=”caps](http://www.uniprot.org/help/DNA&lt;/span&gt;-binding domain</td>
    </tr>
    <tr>
      <td><a href="http://www.uniprot.org/help/np_bind">np_bind</a></td>
      <td>Nucleotide binding</td>
      <td>Nucleotide phosphate binding region</td>
    </tr>
    <tr>
      <td><a href="http://www.uniprot.org/help/region">region</a></td>
      <td>Region</td>
      <td>Region of interest in the sequence</td>
    </tr>
    <tr>
      <td><a href="http://www.uniprot.org/help/coiled">coiled</a></td>
      <td>Coiled coil</td>
      <td>Positions of regions of coiled coil within the protein</td>
    </tr>
    <tr>
      <td><a href="http://www.uniprot.org/help/motif">motif</a></td>
      <td>Motif</td>
      <td>Short (up to 20 amino acids) sequence motif of biological interest</td>
    </tr>
    <tr>
      <td><a href="http://www.uniprot.org/help/act_site">act_site</a></td>
      <td>Active site</td>
      <td>Amino acid(s) directly involved in the activity of an enzyme</td>
    </tr>
    <tr>
      <td><a href="http://www.uniprot.org/help/metal">metal</a></td>
      <td>Metal binding</td>
      <td>Binding site for a metal ion</td>
    </tr>
    <tr>
      <td><a href="http://www.uniprot.org/help/binding">binding</a></td>
      <td>Binding site</td>
      <td>Binding site for any chemical group (co-enzyme, prosthetic group, etc.)</td>
    </tr>
    <tr>
      <td><a href="http://www.uniprot.org/help/site">site</a></td>
      <td>Site</td>
      <td>Any interesting single amino acid site on the sequence</td>
    </tr>
  </tbody>
</table>

<p><strong>Category: Molecule processing</strong></p>

<table>
  <thead>
    <tr>
      <th><strong>ftType</strong></th>
      <th><strong>Label</strong></th>
      <th><strong>Description</strong></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><a href="http://www.uniprot.org/help/init_met">init_met</a></td>
      <td>Initiator methionine</td>
      <td>Cleavage of the initiator methionine</td>
    </tr>
    <tr>
      <td><a href="http://www.uniprot.org/help/signal">signal</a></td>
      <td>Signal</td>
      <td>Sequence targeting proteins to the secretory pathway or periplasmic space</td>
    </tr>
    <tr>
      <td><a href="http://www.uniprot.org/help/transit">transit</a></td>
      <td>Transit peptide</td>
      <td>Extent of a transit peptide for organelle targeting</td>
    </tr>
    <tr>
      <td><a href="http://www.uniprot.org/help/propep">propep</a></td>
      <td>Propeptide</td>
      <td>Part of a protein that is cleaved during maturation or activation</td>
    </tr>
    <tr>
      <td><a href="http://www.uniprot.org/help/chain">chain</a></td>
      <td>Chain</td>
      <td>Extent of a polypeptide chain in the mature protein</td>
    </tr>
    <tr>
      <td><a href="http://www.uniprot.org/help/peptide">peptide</a></td>
      <td>Peptide</td>
      <td>Extent of an active peptide in the mature protein</td>
    </tr>
  </tbody>
</table>

<p><strong>Category: Post translational modifications</strong></p>

<table>
  <thead>
    <tr>
      <th><strong>ftType</strong></th>
      <th><strong>Label</strong></th>
      <th><strong>Description</strong></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><a href="http://www.uniprot.org/help/mod_res">mod_res</a></td>
      <td>Modified residue</td>
      <td>Modified residues excluding lipids, glycans and protein cross-links</td>
    </tr>
    <tr>
      <td><a href="http://www.uniprot.org/help/lipid">lipid</a></td>
      <td>Lipidation</td>
      <td>Covalently attached lipid group(s)</td>
    </tr>
    <tr>
      <td><a href="http://www.uniprot.org/help/carbohyd">carbohyd</a></td>
      <td>Glycosylation</td>
      <td>Covalently attached glycan group(s)</td>
    </tr>
    <tr>
      <td><a href="http://www.uniprot.org/help/disulfid">disulfid</a></td>
      <td>Disulfide bond</td>
      <td>Cysteine residues participating in disulfide bonds</td>
    </tr>
    <tr>
      <td><a href="http://www.uniprot.org/help/crosslnk">crosslnk</a></td>
      <td>Cross-link</td>
      <td>Residues participating in covalent linkage(s) between proteins</td>
    </tr>
  </tbody>
</table>

<p><strong>Category: Sequence information</strong></p>

<table>
  <thead>
    <tr>
      <th><strong>ftType</strong></th>
      <th><strong>Label</strong></th>
      <th><strong>Description</strong></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><a href="http://www.uniprot.org/help/compbias">compbias</a></td>
      <td>Compositional biased</td>
      <td>Region of compositional bias in the protein</td>
    </tr>
    <tr>
      <td><a href="http://www.uniprot.org/help/non_std">non_std</a></td>
      <td>Non-standard residue</td>
      <td>Occurence of non-standard amino acids (selenocysteine and pyrrolysine) in the protein sequence</td>
    </tr>
    <tr>
      <td><a href="http://www.uniprot.org/help/unsure">unsure</a></td>
      <td>Sequence uncertainty</td>
      <td>Regions of uncertainty in the sequence</td>
    </tr>
    <tr>
      <td><a href="http://www.uniprot.org/help/conflict">conflict</a></td>
      <td>Sequence conflict</td>
      <td>Description of sequence discrepancies of unknown origin</td>
    </tr>
    <tr>
      <td><a href="http://www.uniprot.org/help/non_cons">non_cons</a></td>
      <td>Non-adjacent residues</td>
      <td>Indicates that two residues in a sequence are not consecutive</td>
    </tr>
    <tr>
      <td><a href="http://www.uniprot.org/help/non_ter">non_ter</a></td>
      <td>Non-terminal residue</td>
      <td>The sequence is incomplete. Indicate that a residue is not the terminal residue of the complete protein</td>
    </tr>
  </tbody>
</table>

<p><strong>Category: Structural features</strong></p>

<table>
  <thead>
    <tr>
      <th><strong>ftType</strong></th>
      <th><strong>Label</strong></th>
      <th><strong>Description</strong></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><a href="http://www.uniprot.org/help/helix">helix</a></td>
      <td>Helix</td>
      <td>Helical regions within the experimentally determined protein structure</td>
    </tr>
    <tr>
      <td><a href="http://www.uniprot.org/help/turn">turn</a></td>
      <td>Turn</td>
      <td>Turns within the experimentally determined protein structure</td>
    </tr>
    <tr>
      <td><a href="http://www.uniprot.org/help/strand">strand</a></td>
      <td>Beta strand</td>
      <td>Beta strand regions within the experimentally determined protein structure</td>
    </tr>
  </tbody>
</table>

<p><strong>Category: Topology</strong></p>

<table>
  <thead>
    <tr>
      <th><strong>ftType</strong></th>
      <th><strong>Label</strong></th>
      <th><strong>Description</strong></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><a href="http://www.uniprot.org/help/topo_dom">topo_dom</a></td>
      <td>Topological domain</td>
      <td>Location of non-membrane regions of membrane-spanning proteins</td>
    </tr>
    <tr>
      <td><a href="http://www.uniprot.org/help/transmem">transmem</a></td>
      <td>Transmembrane</td>
      <td>Extent of a membrane-spanning region</td>
    </tr>
    <tr>
      <td><a href="http://www.uniprot.org/help/intramem">intramem</a></td>
      <td>Intramembrane</td>
      <td>Extent of a region located in a membrane without crossing it</td>
    </tr>
  </tbody>
</table>

<p><strong>Category: Mutagenesis</strong></p>

<table>
  <thead>
    <tr>
      <th><strong>ftType</strong></th>
      <th><strong>Label</strong></th>
      <th><strong>Description</strong></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><a href="http://www.uniprot.org/help/mutagen">mutagen</a></td>
      <td>Mutagenesis</td>
      <td>Site which has been experimentally altered by mutagenesis</td>
    </tr>
  </tbody>
</table>

<p><strong>Category: Variants</strong></p>

<table>
  <thead>
    <tr>
      <th><strong>ftType</strong></th>
      <th><strong>Label</strong></th>
      <th><strong>Description</strong></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><a href="http://www.uniprot.org/help/variant">variant</a></td>
      <td>Natural variant</td>
      <td>Description of a natural variant of the protein</td>
    </tr>
  </tbody>
</table>

<h2 id="events">Events</h2>
<p>The Protein Feature Viewer triggers five events. In order to listen to those events, you need to get the dispatcher.</p>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="nx">instance</span><span class="p">.</span><span class="nx">getDispatcher</span><span class="p">().</span><span class="nx">on</span><span class="p">(</span><span class="s2">"ready"</span><span class="p">,</span> <span class="kd">function</span><span class="p">(</span><span class="nx">obj</span><span class="p">)</span> <span class="p">{</span>
    <span class="nx">console</span><span class="p">.</span><span class="nx">log</span><span class="p">(</span><span class="s1">'ready'</span><span class="p">);</span>
    <span class="nx">console</span><span class="p">.</span><span class="nx">log</span><span class="p">(</span><span class="nx">obj</span><span class="p">);</span>
<span class="p">});</span>
</code></pre>
</div>

<h3 id="ready">“ready”</h3>
<p>As soon as data has been loaded and visualized, this event will be triggered. The object provided by this event corresponds to all the data retrieved from the data service plus some parsing, it will be ready to be used by the viewer or any other JavaScript component.</p>

<h3 id="featureselected">“featureSelected”</h3>
<p>As soon as a feature has been selected, meaning a click on it has happened, this event will be triggered. The object provided by this event looks like:</p>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="p">{</span>
  <span class="s2">"feature"</span><span class="err">:</span><span class="p">{</span>
    <span class="s2">"type"</span><span class="err">:</span><span class="p">{</span><span class="s2">"name"</span><span class="err">:</span><span class="s2">"REGION"</span><span class="p">,</span><span class="s2">"label"</span><span class="err">:</span><span class="s2">"region of interest"</span><span class="p">},</span>
    <span class="s2">"description"</span><span class="err">:</span><span class="s2">"Heparin-binding"</span><span class="p">,</span>
    <span class="s2">"begin"</span><span class="err">:</span><span class="s2">"96"</span><span class="p">,</span>
    <span class="s2">"end"</span><span class="err">:</span><span class="s2">"110"</span><span class="p">,</span>
    <span class="s2">"internalId"</span><span class="err">:</span><span class="s2">"ft_65"</span>
  <span class="p">},</span>
  <span class="s2">"color"</span><span class="err">:</span><span class="s2">"rgb(179, 62, 0)"</span>
<span class="p">}</span>
</code></pre>
</div>

<h3 id="featuredeselected">“featureDeselected”</h3>
<p>As soon as a feature has been deselected, this event will be triggered. A feature is deselected when it is selected and a clicked on it happens, or when another feature is selected. The object provided by this event looks like the one provided by the event “featureSelected”.</p>

<h3 id="nodata">“noData”</h3>
<p>Triggered when the data failed to be loaded. The object provided by this event will be an <strong>error</strong> return by a jQuery.ajax request.</p>

<h3 id="nofeatures">“noFeatures”</h3>
<p>Triggered when the Uniprot accession has no sequence annotations. The object provided by this event corresponds to the data retrieved from the data service.</p>

<h3 id="notfound">“notFound”</h3>
<p>Triggered when the method <strong>selectFeature</strong> is used but no corresponding feature has been found. The object provided by this event will contain the information passed as parameters to the method, for instance</p>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="p">{</span>
  <span class="s2">"ftType"</span><span class="err">:</span> <span class="s2">"REGION"</span><span class="p">,</span>
  <span class="s2">"begin"</span><span class="err">:</span> <span class="mi">5</span><span class="p">,</span>
  <span class="s2">"end"</span><span class="err">:</span> <span class="mi">27</span>
<span class="p">}</span>
</code></pre>
</div>

<h2 id="contributing">Contributing</h2>
<p>Please submit all issues and pull requests to the <a href="http://github.com/ebi-uniprot/biojs-vis-proteinFeaturesViewer">ebi-uniprot/biojs-vis-proteinFeaturesViewer</a> repository!</p>

<h2 id="support">Support</h2>
<p>If you have any problem or suggestion please open an issue <a href="https://github.com/ebi-uniprot/biojs-vis-proteinFeaturesViewer/issues">here</a>.</p>

<h2 id="license">License</h2>
<p>This software is licensed under the Apache 2 license, quoted below.</p>

<p>Copyright (c) 2014, ebi-uniprot</p>

<p>Licensed under the Apache License, Version 2.0 (the “License”); you may not<br />
use this file except in compliance with the License. You may obtain a copy of<br />
the License at</p>

<div class="highlighter-rouge"><pre class="highlight"><code>http://www.apache.org/licenses/LICENSE-2.0
</code></pre>
</div>

<p>Unless required by applicable law or agreed to in writing, software<br />
distributed under the License is distributed on an “AS IS” BASIS, WITHOUT<br />
WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the<br />
License for the specific language governing permissions and limitations under<br />
the License.</p>

    </div>
</div>
    </section>
</div>

<!-- FOOTER  -->
<div id="footer_wrap" class="outer">
    <footer class="inner">
        <p class="copyright">Biojs-vis-proteinFeaturesViewer maintained by
            <a href="https://github.com/ebi-uniprot">ebi-uniprot</a></p>
        <p>Published with <a href="https://pages.github.com">GitHub Pages</a></p>
    </footer>
</div>



</body>
</html>
