class Covid19
  def self.selected_entries
    [
      {
        name: "7K7P",
        description: "NSP1: Leader protein of 1a and 1ab polyproteins. NSP1 inhibits host cell translation by interacting with the 40S small ribosomal subunit. Slowing down infected host protein production, NSP1 facilitates efficient viral protein translation and evasion from host immune response, thus helping the virus to proliferate.",
        image_url: "https://www.ebi.ac.uk/pdbe/static/entry/7k7p_deposited_chain_front_image-800x800.png",
        query_url: "/ws/viewer/#/7k7p"
      },
      {
        name: "7MSW + EMD-23970",
        description: "NSP2: Although its function is still unknown, NSP2 has been involved in several viral processes. Its highly-conserved Zn2+ binding site suggests that nsp2 binds RNA. It also seems to interact with host infected cell endosomes through cytoskeletal elements and with modulators of translation. ",
        image_url: "images/entries/7msw.png",
        query_url: "/ws/viewer/#/7msw+EMD-23970"
      },
      {
        name: "7TI9",
        description: "Ubiquitin-like domain 1 of Nsp3.",
        image_url: "https://www.ebi.ac.uk/pdbe/static/entry/7ti9_deposited_chain_front_image-800x800.png",
        query_url: "/ws/viewer/#/7ti9"
      },
      {
        name: "5S73",
        description: "MacroDomain I of NSP3 (Mac1): Domain that recognizes and hydrolizes the ADP-ribose label from proteins associated with the host innate immune response.",
        image_url: "https://www.ebi.ac.uk/pdbe/static/entry/5s73_deposited_chain_front_image-800x800.png",
        query_url: "/ws/viewer/#/5s73"
      },
      {
        name: "7THH",
        description: "SARS-unique domain C of NSP3 (DPUP).",
        image_url: "https://www.ebi.ac.uk/pdbe/static/entry/7thh_deposited_chain_front_image-800x800.png",
        query_url: "/ws/viewer/#/7thh"
      },
      {
        name: "7NFV",
        description: "PL-Pro: Domain Papain-like proteinase of NSP3. PL-Pro processes proteolytically three cleavage sites located at the N-terminus of the polyproteins 1a and 1ab. In addition, PL-Pro possesses a deubiquitinating/deISGylating activity and processes polyubiquitin chains from cellular substrates. It participates together with NSP4 in the assembly of virally-induced cytoplasmic double-membrane vesicles necessary for viral replication. It also antagonizes innate immune induction of type I interferon by blocking phosphorylation, dimerization and subsequent nuclear translocation of host IRF3. It prevents host cell NF-kappa-B signaling as well.",
        image_url: "https://www.ebi.ac.uk/pdbe/static/entry/7nfv_deposited_chain_front_image-800x800.png",
        query_url: "/ws/viewer/#/7nfv"
      },
      {
        name: "7LGO",
        description: "NAB: Nucleic acid binding domain of NSP3.",
        image_url: "https://www.ebi.ac.uk/pdbe/static/entry/7lgo_deposited_chain_front_image-800x800.png",
        query_url: "/ws/viewer/#/7lgo"
      },
      {
        name: "7RQG",
        description: "Y3 domain of Nsp3.",
        image_url: "https://www.ebi.ac.uk/pdbe/static/entry/7rqg_deposited_chain_front_image-800x800.png",
        query_url: "/ws/viewer/#/7rqg"
      },
      {
        name: "5R8T",
        description: "NSP5: Protease that processes proteolytically the C-terminus of the polyproteins 1a and 1ab at seven and twelve sites, respectively.",
        image_url: "https://www.ebi.ac.uk/pdbe/static/entry/5r8t_deposited_chain_front_image-800x800.png",
        query_url: "/ws/viewer/#/5r8t"
      },
      {
        name: "7C2K + EMD-30275",
        description: "NSP7: As cofactor of NSP12, NSP7 contributes to accomodate the template-product RNA, together with NSP8, in the RNA-dependent RNA polymerase (RdRp) catalytic complex.",
        image_url: "images/entries/7c2k.png",
        query_url: "/ws/viewer/#/7c2k+EMD-30275"
      },
      {
        name: "6WXD",
        description: "NSP9: Able to bind ssRNA, NSP9 forms a covalent RNA-protein intermediate. This RNA is first transferred by the NiRAN domain of NSP12 to NSP9 . Then, the NiRAN NSP12 domain transfers the RNA to GDP to start the building process of the RNA cap structure. Bound to the NSP12 NiRAN domain, NSP9 inhibits the guanylyl tranferase enzymatic activity of NSP12.",
        image_url: "https://www.ebi.ac.uk/pdbe/static/entry/6wxd_deposited_chain_front_image-800x800.png",
        query_url: "/ws/viewer/#/6wxd"
      },
      {
        name: "6ZPE",
        description: "NSP10: Stimulating both NSP14 exoribonuclease and NSP16 methyltransferase activities, NSP10 plays an essential role in viral mRNA proofreading and cap methylation.",
        image_url: "https://www.ebi.ac.uk/pdbe/static/entry/6zpe_deposited_chain_front_image-800x800.png",
        query_url: "/ws/viewer/#/6zpe"
      },
      {
        name: "5ROB",
        description: "NSP13: Its helicase RNA duplex-unwinding activity contributes to elongate the RNA product. Additionally, its fosfatase activity catalizes the first step of the mRNA cap synthesis.",
        image_url: "https://www.ebi.ac.uk/pdbe/static/entry/5rob_deposited_chain_front_image-800x800.png",
        query_url: "/ws/viewer/#/5rob"
      },
      {
        name: "5SMK",
        description: "NSP14: Enzyme with two different catalytic activities: Exoribonuclease activity in its N-terminal end, and N7-methyltransferase activity in its C-terminal end. Its proofreading exoribonuclease lowers viral sensitivity to nucleoside analogs.",
        image_url: "https://www.ebi.ac.uk/pdbe/static/entry/5smk_deposited_chain_front_image-800x800.png",
        query_url: "/ws/viewer/#/5smk"
      },
      {
        name: "5SBF",
        description: "NSP15: As uridylate-specific enzyme, NSP15 hydrolizes both ssRNA and dsRNA, probably to avoid trigger the host inmune response.",
        image_url: "https://www.ebi.ac.uk/pdbe/static/entry/5sbf_deposited_chain_front_image-800x800.png",
        query_url: "/ws/viewer/#/5sbf"
      },
      {
        name: "7L6T",
        description: "NSP16: Methyltransferase that catalyzes the last reaction to build the viral mRNA cap, the methylation of the ribose 2'-OH.",
        image_url: "https://www.ebi.ac.uk/pdbe/static/entry/7l6t_deposited_chain_front_image-800x800.png",
        query_url: "/ws/viewer/#/7l6t"
      },
      {
        name: "6VSB + EMD-21375",
        description: "Spike: Decorating surface glycoprotein that protrudes from the viral outer membrane. Several conformational changes allow its interaction with the host receptor ACE2 and membrane fusion, thus initiating the infection.",
        image_url: "images/entries/6vsb.png",
        query_url: "/ws/viewer/#/6vsb+EMD-21375"
      },
      {
        name: "6XDC + EMD-22136",
        description: "ORF3: As a viroporin, ORF3 forms cellular dimeric nonselective Ca2+ permeable cation channels that become active by oligomerizing into tetramers or higher-order oligomers. This channel activity in cells could be relevant to promote viral maturation through inhibition of autophagy and disruption of lysosomes.",
        image_url: "images/entries/6xcd.png",
        query_url: "/ws/viewer/#/6xdc+EMD-22136"
      },
      {
        name: "7NTK",
        description: "Protein E: Envelope small membrane protein. As a viroporin, it self-assembles in host membranes forming pentameric protein pores that can affect the integrity of the lipid bilayer or disrupt the membrane potential, thus facilitating the release of viral particles from host cells.",
        image_url: "https://www.ebi.ac.uk/pdbe/static/entry/7ntk_deposited_chain_front_image-800x800.png",
        query_url: "/ws/viewer/#/7ntk"
      },
      {
        name: "7VPH",
        description: "ORF6: Although it is ot necessary for viral invasion and replication, ORF6 plays a relevant role in immune evasion. It directly interacts with the host complex Rae1-Nup98 preventing the nuclear export of host mRNA, antagonizing interferon signaling.",
        image_url: "https://www.ebi.ac.uk/pdbe/static/entry/7vph_deposited_chain_front_image-800x800.png",
        query_url: "/ws/viewer/#/7vph"
      },
      {
        name: "7CI3",
        description: "ORF7a: Although dispensable for virus replication in cell culture, ORF7a acts as a immunomodulator factor that binds to immune cells and triggers inflammatory responses.",
        image_url: "https://www.ebi.ac.uk/pdbe/static/entry/7ci3_deposited_chain_front_image-800x800.png",
        query_url: "/ws/viewer/#/7ci3"
      },
      {
        name: "7F5F",
        description: "ORF8: With a core fold similar to ORF7a, the structure of ORF8 reveals two novel dimer interfaces that allow forming host large-scale assemblies, potentially mediating immune suppression and evasion activities.",
        image_url: "https://www.ebi.ac.uk/pdbe/static/entry/7jtl_deposited_chain_front_image-800x800.png",
        query_url: "/ws/viewer/#/7f5f"
      },
      {
        name: "7CDZ",
        description: "Nucleoprotein N-terminal domain: The nucleoprotein (N protein) is involved in viral assembly, replication and host immune response regulation. Its N-terminal domain (N-NTD) is responsible for RNA binding and packaging of the positive strand viral genome RNA into a helical ribonucleocapsid (RNP).",
        image_url: "https://www.ebi.ac.uk/pdbe/static/entry/7cdz_deposited_chain_front_image-800x800.png",
        query_url: "/ws/viewer/#/7cdz"
      },
      {
        name: "7CE0",
        description: "Nucleoprotein C-terminal domain: The nucleoprotein (N protein) is involved in viral assembly, replication and host immune response regulation. Its C-terminal domain (N-CTD) is responsible for RNA binding and packaging of the positive strand viral genome RNA into a helical ribonucleocapsid (RNP).",
        image_url: "https://www.ebi.ac.uk/pdbe/static/entry/7ce0_deposited_chain_front_image-800x800.png",
        query_url: "/ws/viewer/#/7ce0"
      },
      {
        name: "7KDT + EMD-22829",
        description: "ORF9b: The gene encoding this small protein overlaps the N gene. ORF9b protein interacts and blocks the substrate binding site of the host Tom70, import receptor of the translocase from the mitochondrial outer membrane (TOM) complex, involved in the activation of the mitochondrial antiviral signaling leading to apoptosis upon viral infection. In addition, ORF9b may modulate interferon and apoptosis signaling avoiding the interaction between Tom70 and Hsp70 and Hsp90 chaperones.",
        image_url: "images/entries/7kdt.png",
        query_url: "/ws/viewer/#/7kdt+EMD-22829"
      },
    ]
  end

  def self.proteins_data
    load_data
  end

  private

  def self.get_pdb_redo_links(title, style, pdb_key, hash, keys)
    entries0 = hash && keys ? (hash.dig(*keys) || []) : []
    entries = entries0.is_a?(Hash) ? [entries0] : entries0

    with_indexes(entries, title).map do |entry, title_indexed|
      name, = entry.values_at("name")
      external_url = "https://pdb-redo.eu/db/#{name}"
      {
        title: title_indexed,
        name: name,
        style: style,
        # query_url: "/ws/viewer/#/PDB-REDO-#{pdb_key}&viewer_type=ngl&button=#query",
        query_url: "/pdb_redo/#{pdb_key}",
        external_url: external_url,
        check: external_url,
      }
    end
  end

  def self.with_indexes(entries, title)
    append_index = entries.size > 1
    entries.map.with_index(1) { |entry, idx| [entry, append_index ? "#{title} (#{idx})" : title] }
  end

  def self.get_isolde_links(title, style, pdb_key, hash, keys)
    return [] unless hash && keys
    entries = hash.dig(*keys) || []

    with_indexes(entries, title).map do |entry, title_indexed|
      entry_name, uuid = entry.values_at("name", "uuid")
      {
        title: title_indexed,
        name: entry_name,
        style: style,
        query_url: "/isolde/#{pdb_key}/#{entry_name}",
      }
    end
  end

  def self.get_refmac_links(title, style, pdb_key, hash, keys)
    return [] unless hash && keys
    entries = hash.dig(*keys) || []

    with_indexes(entries, title).map do |entry, title_indexed|
      entry_name, uuid = entry.values_at("name", "uuid")
      {
        title: title_indexed,
        name: entry_name,
        style: style,
        query_url: "/refmac/#{pdb_key}/#{entry_name}",
      }
    end
  end

  def self.get_related_keys(data)
    case
    when data.is_a?(Array)
      data
    when data.is_a?(Hash)
      data.keys
    else
      []
    end
  end

  def self.to_hash(entries)
    entries.is_a?(Hash) ? entries : entries.map { |key| [key, {}] }.to_h.sort
  end

  def self.parse_pdb(protein, keys)
    entries = protein.dig(*keys) || []
    items = []
    pockets = {}
    to_hash(entries).each do |pdb_key, pdb_hash|
      items.push({
        name: pdb_key,
        description: pdb_hash["description"],
        query_url: "/ws/viewer/#/#{pdb_key}&viewer_type=ngl&button=#query",
        image_url: "https://www.ebi.ac.uk/pdbe/static/entry/#{pdb_key}_deposited_chain_front_image-800x800.png",
        external: {text: "EBI", url: "https://www.ebi.ac.uk/pdbe/entry/pdb/#{pdb_key}"},
        api: "https://www.ebi.ac.uk/pdbe/api/pdb/entry/summary/#{pdb_key}",
        type: 'pdb',
        related: get_related_keys(pdb_hash["emdbs"]),
        pockets: pdb_hash.key?("pockets") ? pdb_hash["pockets"].map { |pocket| pocket['id']} : [],
        experiment: keys.last,
        links: [
          *get_pdb_redo_links("PDB-Redo", :turq, pdb_key, pdb_hash, ["validation", "pdb-redo"]),
          *get_isolde_links("Isolde", :cyan, pdb_key, pdb_hash, ["validation", "isolde"]),
          *get_refmac_links("Refmac", :blue_gray, pdb_key, pdb_hash, ["validation", "refmac"]),
        ],
      })

      if pdb_hash.key?("pockets")
        pdb_hash["pockets"].each do |pocket|
          id = pocket['id']
          bindingSiteScore = pocket['bindingSiteScore']
          pockets[id] = bindingSiteScore
        end
      end
    end
    return items.compact, pockets
  end

  def self.parse_emdb(protein, keys)
    entries = protein.dig(*keys) || []
    to_hash(entries).map do |emdb_key, emdb_hash|
      code = emdb_key.split("-")[1]
      {
        name: emdb_key,
        description: emdb_hash["description"],
        query_url: "/ws/viewer/#/#{emdb_key}&viewer_type=ngl&button=#query",
        image_url: "https://www.ebi.ac.uk/pdbe/static/entry/#{emdb_key}/400_#{code}.gif",
        related: get_related_keys(emdb_hash["pdbs"] || emdb_hash["pdb"]),
        external: {text: "EBI", url: "https://www.ebi.ac.uk/pdbe/entry/emdb/#{emdb_key}"},
        api: "https://www.ebi.ac.uk/pdbe/api/emdb/entry/summary/#{emdb_key}",
        type: 'emdb',
        links: [],
      }
    end
  end

  def self.parse_swiss_model(protein, keys)
    entries = protein.dig(*keys) || []
    entries.map do |entry|
      project, model, description = entry.values_at("project", "model", "description")
      {
        name: "#{project}-#{model}",
        description: ["project: #{project}", "model: #{model}", description].compact.join(" | "),
        # query_url: "/ws/viewer/#/SWISSMODEL-#{protein['uniprotAccession'][0]}-#{project}-#{model}&viewer_type=ngl&button=#query",
        query_url: "/models/#{protein['uniprotAccession'][0]}/#{keys[1]}/#{project}-#{model}",
        image_url: "https://swissmodel.expasy.org/interactive/#{project}/models/#{model}.png",
        external: {text: "SWISS-MODEL", url: "https://swissmodel.expasy.org/interactive/#{project}/models/#{model}"},
        links: [],
      }
    end
  end

  def self.parse_alphafold(protein, keys)
    entries = protein.dig(*keys) || []
    entries.map do |entry|
      name, uuid, description = entry.values_at("name", "uuid", "description")
      {
        name: name,
        description: description,
        query_url: "/models/#{protein['uniprotAccession'][0]}/#{keys[1]}/#{name}",
        image_url: "/AlphaFold_logo.png",
        external: {text: "AlphaFold", url: "https://deepmind.com/research/open-source/computational-predictions-of-protein-structures-associated-with-COVID-19"},
        links: [],
      }
    end
  end

  def self.parse_bsm_arc(protein, keys)
    entries = protein.dig(*keys) || []
    entries.map do |entry|
      model, = entry.values_at("model")
      {
        name: "#{model}",
        query_url: "/models/#{protein['uniprotAccession'][0]}/#{keys[1]}/#{model}",
        external: {text: "BSM-Arc", url: "https://bsma.pdbj.org/entry/15"},
        links: [],
      }
    end
  end

  def self.parse_db_with_experiments(protein, keys)
    entries = protein.dig(*keys) || []

    experiments = []
    items = []
    pockets = {}

    entries.each do |key,value|
      if key == 'EMDB'
        items = items + parse_emdb(protein, [*keys, "EMDB"])
      else
        new_items, pdb_pockets = parse_pdb(protein, [*keys, key])
        if new_items.size > 0 
          items = items + new_items
          experiments.push(key)
          pockets.merge!(pdb_pockets)
        end
      end
    end
    return items, experiments, pockets
    
  end

  def self.parse_db(protein, keys)
    parse_pdb(protein, [*keys, "PDB"]) + parse_emdb(protein, [*keys, "EMDB"])
  end

  def self.load_data
    json_path = File.join(__dir__, "../data/cv-data.json")
    data = JSON.parse(open(json_path).read)
    get_proteins_data(data["SARS-CoV-2 Proteome"])
  end

  def self.card(name, items, experiments = [], pockets = [])
    items.present? ? {name: name, items: items, experiments: experiments, pockets: pockets, subsections: []} : nil
  end

  def self.card_wrapper(name, subsections)
    subsections2 = subsections.compact.map { |subsection| subsection.merge(parent: name) }
    subsections2.size > 0 ? {name: name, items: [], experiments: [], pockets: [], subsections: subsections2} : nil
  end

  def self.get_relations(proteins)
    items = {}
    relations_base = proteins.each do |protein|
      protein[:sections].each do |section|
        section[:items].each do |item|
          key = item[:name] + "-" + protein[:name] + "-" + section[:name].parameterize
          items[key] = {protein: protein[:name], experiment: item[:experiment], pockets: item[:pockets]}
        end
        subsection_items = section[:subsections].flat_map do |subsection|
          subsection[:items].each do |item|
            key = item[:name] + "-" + protein[:name] + "-" + subsection[:name].parameterize
            items[key] = {protein: protein[:name], experiment: item[:experiment], pockets: item[:pockets]}
          end
        end
      end
    end

    return items
  end

  def self.get_proteins_data(proteins_raw)
    proteins = proteins_raw.map do |name, protein|
      {
        name: name,
        names: protein["names"],
        description: protein["description"],
        polyproteins: protein["uniprotAccession"],
        sections: [
          card("PDB", *parse_pdb(protein, ["PDB"])),
          card("EMDB", parse_emdb(protein, ["EMDB"])),
          card_wrapper("Interactions", [
            card("Protein-Protein Interactions", *parse_db_with_experiments(protein, ["Interactions", "P-P-Interactions"])),
            card("Ligands", *parse_db_with_experiments(protein, ["Interactions", "Ligands"])),
          ]),
          card_wrapper("Related", [
            card("SARS-CoV", *parse_db_with_experiments(protein, ["Related", "SARS-CoV"])),
            card("Other", *parse_db_with_experiments(protein, ["Related", "OtherRelated"]))
          ]),
          card_wrapper("Computational Models", [
            card("Swiss Model", parse_swiss_model(protein, ["CompModels", "swiss-model"])),
            card("AlphaFold", parse_alphafold(protein, ["CompModels", "AlphaFold"])),
            card("BSM-Arc", parse_bsm_arc(protein, ["CompModels", "BSM-Arc"])),
          ]),

        ].compact,
      }
    end

    {proteins: proteins, relations: get_relations(proteins)}
  end
end
