class Covid19
  def self.selected_entries
    [
      {
        name: "7K7P",
        description: "NSP1. Leader protein. Host translation inhibitor. Part of Replicase polyprotein 1a and 1ab. Inhibits host translation by interacting with the 40S ribosomal subunit. The nsp1-40S ribosome complex further induces an endonucleolytic cleavage near the 5'UTR of host mRNAs, targeting them for degradation. Viral mRNAs are not susceptible to nsp1-mediated endonucleolytic RNA cleavage thanks to the presence of a 5'-end leader sequence and are therefore protected from degradation. By suppressing host gene expression, nsp1 facilitates efficient viral gene expression in infected cells and evasion from host immune response.",
        image_url: "https://www.ebi.ac.uk/pdbe/static/entry/7k7p_deposited_chain_front_image-200x200.png",
        query_url: "/?queryId=7k7p"
      },
      {
        name: "7MSW",
        description: "NSP2. Non-structural protein 2. May play a role in the modulation of host cell survival signaling pathway by interacting with host PHB and PHB2. Indeed, these two proteins play a role in maintaining the functional integrity of the mitochondria and protecting cells from various stresses.",
        image_url: "https://www.ebi.ac.uk/pdbe/static/entry/7msw_deposited_chain_front_image-200x200.png",
        query_url: "/?queryId=7msw"
      },
      {
        name: "EMD-23970",
        description: "NSP2. Non-structural protein 2. May play a role in the modulation of host cell survival signaling pathway by interacting with host PHB and PHB2. Indeed, these two proteins play a role in maintaining the functional integrity of the mitochondria and protecting cells from various stresses.",
        image_url: "https://www.ebi.ac.uk/pdbe/static/entry/EMD-23970/400_23970.gif",
        query_url: "/?queryId=emd-23970"
      },
      {
        name: "7TI9",
        description: "Ubiquitin-like domain 1 of Nsp3.",
        image_url: "https://www.ebi.ac.uk/pdbe/static/entry/7ti9_deposited_chain_front_image-200x200.png",
        query_url: "/?queryId=7ti9"
      },
      {
        name: "7CZ4",
        description: "Macrodomain of Nsp3 (Mac1). Function: It recognizes and hydrolizes the ADP-ribose label from proteins associated with the host innate immune response.",
        image_url: "https://www.ebi.ac.uk/pdbe/static/entry/7cz4_deposited_chain_front_image-200x200.png",
        query_url: "/?queryId=7cz4"
      },
      {
        name: "7THH",
        description: "SARS-unique domain C of Nsp3 (renamed as domain preceding Ubl-2 and PL-Pro, DPUP https://doi.org/10.15252/embj.2019102277).",
        image_url: "https://www.ebi.ac.uk/pdbe/static/entry/7thh_deposited_chain_front_image-200x200.png",
        query_url: "/?queryId=7thh"
      },
      {
        name: "7THH",
        description: "Ubiquitin-like domain 2 of Nsp3.",
        image_url: "https://www.ebi.ac.uk/pdbe/static/entry/7thh_deposited_chain_front_image-200x200.png",
        query_url: "/?queryId=7thh"
      },
      {
        name: "7NFV",
        description: "NSP3. PL-Pro. Papain-like proteinase. Part of Replicase polyprotein 1a and 1ab. Responsible for the cleavages located at the N-terminus of the replicase polyprotein. In addition, PL-PRO possesses a deubiquitinating/deISGylating activity and processes both 'Lys-48'- and 'Lys-63'-linked polyubiquitin chains from cellular substrates. Participates together with nsp4 in the assembly of virally-induced cytoplasmic double-membrane vesicles necessary for viral replication. Antagonizes innate immune induction of type I interferon by blocking the phosphorylation, dimerization and subsequent nuclear translocation of host IRF3. Prevents also host NF-kappa-B signaling.",
        image_url: "https://www.ebi.ac.uk/pdbe/static/entry/7nfv_deposited_chain_front_image-200x200.png",
        query_url: "/?queryId=7nfv"
      },
      {
        name: "7LGO",
        description: "Nucleic acid binding domain of Nsp3.",
        image_url: "https://www.ebi.ac.uk/pdbe/static/entry/7lgo_deposited_chain_front_image-200x200.png",
        query_url: "/?queryId=7lgo"
      },
      {
        name: "7RQG",
        description: "Y3 domain of Nsp3.",
        image_url: "https://www.ebi.ac.uk/pdbe/static/entry/7rqg_deposited_chain_front_image-200x200.png",
        query_url: "/?queryId=7rqg"
      },
      {
        name: "5R8T",
        description: "NSP5. Main Protease. MPro. 3C-like protease. 3CL-Pro. Part of Replicase polyprotein 1a and 1ab. Cleaves the C-terminus of replicase polyprotein at 11 sites.",
        image_url: "https://www.ebi.ac.uk/pdbe/static/entry/5r8t_deposited_chain_front_image-200x200.png",
        query_url: "/?queryId=5r8t"
      },
      {
        name: "7C2K",
        description: "NSP7. Non-structural protein 7. Part of Replicase polyprotein 1a and 1ab. Forms a hexadecamer with nsp8 (8 subunits of each) that may participate in viral replication by acting as a primase. Alternatively, may synthesize substantially longer products than oligonucleotide primers.",
        image_url: "https://www.ebi.ac.uk/pdbe/static/entry/7c2k_deposited_chain_front_image-200x200.png",
        query_url: "/?queryId=7c2k"
      },
      {
        name: "EMD-30275",
        description: "NSP7. Non-structural protein 7. Part of Replicase polyprotein 1a and 1ab. Forms a hexadecamer with nsp8 (8 subunits of each) that may participate in viral replication by acting as a primase. Alternatively, may synthesize substantially longer products than oligonucleotide primers.",
        image_url: "https://www.ebi.ac.uk/pdbe/static/entry/EMD-30275/400_30275.gif",
        query_url: "/?queryId=emd-30275"
      },
      {
        name: "7C2K",
        description: "NSP8. Non-structural protein 8. Part of Replicase polyprotein 1a and 1ab. Forms a hexadecamer with nsp7 (8 subunits of each) that may participate in viral replication by acting as a primase. Alternatively, may synthesize substantially longer products than oligonucleotide primers.",
        image_url: "https://www.ebi.ac.uk/pdbe/static/entry/7c2k_deposited_chain_front_image-200x200.png",
        query_url: "/?queryId=7c2k"
      },
      {
        name: "EMD-30275",
        description: "NSP8. Non-structural protein 8. Part of Replicase polyprotein 1a and 1ab. Forms a hexadecamer with nsp7 (8 subunits of each) that may participate in viral replication by acting as a primase. Alternatively, may synthesize substantially longer products than oligonucleotide primers.",
        image_url: "https://www.ebi.ac.uk/pdbe/static/entry/EMD-30275/400_30275.gif",
        query_url: "/?queryId=emd-30275"
      },
      {
        name: "7BWQ",
        description: "NSP9. Non-structural protein 9. ssRNA binding protein. Part of Replicase polyprotein 1a and 1ab. May participate in viral replication by acting as a ssRNA-binding protein.",
        image_url: "https://www.ebi.ac.uk/pdbe/static/entry/7bwq_deposited_chain_front_image-200x200.png",
        query_url: "/?queryId=7bwq"
      },
      {
        name: "6ZCT",
        description: "NSP10. Non-structural protein 10. Part of Replicase polyprotein 1a and 1ab. Plays a pivotal role in viral transcription by stimulating both nsp14 3'-5' exoribonuclease and nsp16 2'-O-methyltransferase activities. Therefore plays an essential role in viral mRNAs cap methylation.",
        image_url: "https://www.ebi.ac.uk/pdbe/static/entry/6zct_deposited_chain_front_image-200x200.png",
        query_url: "/?queryId=6zct"
      },
      {
        name: "7C2K",
        description: "NSP12. RNA-directed RNA polymerase. RNA-dependent RNA polymerase. Part of Replicase polyprotein 1ab. Responsible for replication and transcription of the viral RNA genome.",
        image_url: "https://www.ebi.ac.uk/pdbe/static/entry/7c2k_deposited_chain_front_image-200x200.png",
        query_url: "/?queryId=7c2k"
      },
      {
        name: "EMD-30275",
        description: "NSP12. RNA-directed RNA polymerase. RNA-dependent RNA polymerase. Part of Replicase polyprotein 1ab. Responsible for replication and transcription of the viral RNA genome.",
        image_url: "https://www.ebi.ac.uk/pdbe/static/entry/EMD-30275/400_30275.gif",
        query_url: "/?queryId=emd-30275"
      },
      {
        name: "5ROB",
        description: "NSP13 | Helicase. Part of Replicase polyprotein 1ab. Multi-functional protein with a zinc-binding domain in N-terminus displaying RNA and DNA duplex-unwinding activities with 5' to 3' polarity. Activity of helicase is dependent on magnesium.",
        image_url: "https://www.ebi.ac.uk/pdbe/static/entry/5rob_deposited_chain_front_image-200x200.png",
        query_url: "/?queryId=5rob"
      },
      {
        name: "5SMK",
        description: "NSP14. Proofreading exoribonuclease. 3'-5' exoribonuclease. Guanine-N7 methyltransferase. ExoN/nsp14. Part of Replicase polyprotein 1ab. Enzyme possessing two different activities: an exoribonuclease activity acting on both ssRNA and dsRNA in a 3' to 5' direction and a N7-guanine methyltransferase activity. Acts as a proofreading exoribonuclease for RNA replication, thereby lowering The sensitivity of the virus to RNA mutagens.",
        image_url: "https://www.ebi.ac.uk/pdbe/static/entry/5smk_deposited_chain_front_image-200x200.png",
        query_url: "/?queryId=5smk"
      },
      {
        name: "5SBF",
        description: "NSP15. Non-structural protein 15. Endoribonuclease. Uridylate-specific endoribonuclease. Endornase. NendoU/nsp15. Part of Replicase polyprotein 1ab. Mn(2+)-dependent, uridylate-specific enzyme, which leaves 2'-3'-cyclic phosphates 5' to the cleaved bond.",
        image_url: "https://www.ebi.ac.uk/pdbe/static/entry/5sbf_deposited_chain_front_image-200x200.png",
        query_url: "/?queryId=5sbf"
      },
      {
        name: "7L6T",
        description: "NSP16. Methyltransferase. 2'-O-methyltransferase. 2'-O-ribose methyltransferase. Part of Replicase polyprotein 1ab. Methyltransferase that mediates mRNA cap 2'-O-ribose methylation to the 5'-cap structure of viral mRNAs. N7-methyl guanosine cap is a prerequisite for binding of nsp16. Therefore plays an essential role in viral mRNAs cap methylation which is essential to evade immune system.",
        image_url: "https://www.ebi.ac.uk/pdbe/static/entry/7l6t_deposited_chain_front_image-200x200.png",
        query_url: "/?queryId=7l6t"
      },
      {
        name: "6VSB",
        description: "Spike. Protein S. Spike glycoprotein. Surface Glycoprotein. Spike protein, trimeric complex S1-S2-S2': Attaches the virion to the cell membrane by interacting with host receptor, initiating the infection. Binding to human ACE2 receptor and internalization of the virus into the endosomes of the host cell induces conformational changes in the Spike glycoprotein. Uses also human TMPRSS2 for priming in human lung cells which is an essential step for viral entry. Proteolysis by cathepsin CTSL may unmask the fusion peptide of S2 and activate membranes fusion within endosomes.",
        image_url: "https://www.ebi.ac.uk/pdbe/static/entry/6vsb_deposited_chain_front_image-200x200.png",
        query_url: "/?queryId=6vsb"
      },
      {
        name: "EMD-21375",
        description: "Spike. Protein S. Spike glycoprotein. Surface Glycoprotein. Spike protein, trimeric complex S1-S2-S2': Attaches the virion to the cell membrane by interacting with host receptor, initiating the infection. Binding to human ACE2 receptor and internalization of the virus into the endosomes of the host cell induces conformational changes in the Spike glycoprotein. Uses also human TMPRSS2 for priming in human lung cells which is an essential step for viral entry. Proteolysis by cathepsin CTSL may unmask the fusion peptide of S2 and activate membranes fusion within endosomes.",
        image_url: "https://www.ebi.ac.uk/pdbe/static/entry/EMD-21375/400_21375.gif",
        query_url: "/?queryId=emd-21375"
      },
      {
        name: "6XCD",
        description: "ORF3. Protein 3. Accessory protein 3a. Protein U274. Protein X1. Forms homotetrameric potassium sensitive ion channels (viroporin) and may modulate virus release. Up-regulates expression of fibrinogen subunits FGA, FGB and FGG in host lung epithelial cells. Induces apoptosis in cell culture. Downregulates the type 1 interferon receptor by inducing serine phosphorylation within the IFN alpha-receptor subunit 1 (IFNAR1) degradation motif and increasing IFNAR1 ubiquitination.",
        image_url: "https://www.ebi.ac.uk/pdbe/static/entry/6xcd_deposited_chain_front_image-200x200.png",
        query_url: "/?queryId=6xcd"
      },
      {
        name: "EMD-22136",
        description: "ORF3. Protein 3. Accessory protein 3a. Protein U274. Protein X1. Forms homotetrameric potassium sensitive ion channels (viroporin) and may modulate virus release. Up-regulates expression of fibrinogen subunits FGA, FGB and FGG in host lung epithelial cells. Induces apoptosis in cell culture. Downregulates the type 1 interferon receptor by inducing serine phosphorylation within the IFN alpha-receptor subunit 1 (IFNAR1) degradation motif and increasing IFNAR1 ubiquitination.",
        image_url: "https://www.ebi.ac.uk/pdbe/static/entry/EMD-22136/400_22136.gif",
        query_url: "/?queryId=emd-22136"
      },
      {
        name: "7NTK",
        description: "E. Envelope small membrane protein. Plays a central role in virus morphogenesis and assembly. Acts as a viroporin and self-assembles in host membranes forming pentameric protein-lipid pores that allow ion transport. Also plays a role in the induction of apoptosis.",
        image_url: "https://www.ebi.ac.uk/pdbe/static/entry/7ntk_deposited_chain_front_image-200x200.png",
        query_url: "/?queryId=7ntk"
      },
      {
        name: "7VPH",
        description: "ORF6. Accessory Protein 6. Accessory Factor 6. Disrupts cell nuclear import complex formation by tethering karyopherin alpha 2 and karyopherin beta 1 to the membrane. Retention of import factors at the ER/Golgi membrane leads to a loss of transport into the nucleus. Thereby prevents STAT1 nuclear translocation in response to interferon signaling, thus blocking the expression of interferon stimulated genes (ISGs) that display multiple antiviral activities.",
        image_url: "https://www.ebi.ac.uk/pdbe/static/entry/7vph_deposited_chain_front_image-200x200.png",
        query_url: "/?queryId=7vph"
      },
      {
        name: "7CI3",
        description: "ORF7a Protein. Non-structural protein which is dispensable for virus replication in cell culture.",
        image_url: "https://www.ebi.ac.uk/pdbe/static/entry/7ci3_deposited_chain_front_image-200x200.png",
        query_url: "/?queryId=7ci3"
      },
      {
        name: "7JTL",
        description: "ORF8. Non-structural protein 8. May play a role in host-virus interaction. Simnilar to some Bat coronavirus ns8 genes, but is entirely different from SARS ns8a or Ns8b.",
        image_url: "https://www.ebi.ac.uk/pdbe/static/entry/7jtl_deposited_chain_front_image-200x200.png",
        query_url: "/?queryId=7jtl"
      },
      {
        name: "7CDZ",
        description: "Protein N. Nucleoprotein. Nucleocapsid protein. Packages the positive strand viral genome RNA into a helical ribonucleocapsid (RNP) and plays a fundamental role during virion assembly through its interactions with the viral genome and membrane protein M. Plays an important role in enhancing the efficiency of subgenomic viral RNA transcription as well as viral replication (By similarity). May modulate transforming growth factor-beta signaling by binding host smad3 (By similarity).",
        image_url: "https://www.ebi.ac.uk/pdbe/static/entry/7cdz_deposited_chain_front_image-200x200.png",
        query_url: "/?queryId=7cdz"
      },
      {
        name: "7CE0",
        description: "",
        image_url: "https://www.ebi.ac.uk/pdbe/static/entry/7ce0_deposited_chain_front_image-200x200.png",
        query_url: "/?queryId=7ce0"
      },
      {
        name: "7KDT",
        description: "ORF9b. Protein 9b. The gene encoding this protein is included within the N gene (alternative ORF). Forms homotetrameric potassium sensitive ion channels (viroporin) and may modulate virus release. Up-regulates expression of fibrinogen subunits FGA, FGB and FGG in host lung epithelial cells. Induces apoptosis in cell culture. Downregulates the type 1 interferon receptor by inducing serine phosphorylation within the IFN alpha-receptor subunit 1 (IFNAR1) degradation motif and increasing IFNAR1 ubiquitination.",
        image_url: "https://www.ebi.ac.uk/pdbe/static/entry/7kdt_deposited_chain_front_image-200x200.png",
        query_url: "/?queryId=7kdt"
      },
      {
        name: "EMD-22829",
        description: "ORF9b. Protein 9b. The gene encoding this protein is included within the N gene (alternative ORF). Forms homotetrameric potassium sensitive ion channels (viroporin) and may modulate virus release. Up-regulates expression of fibrinogen subunits FGA, FGB and FGG in host lung epithelial cells. Induces apoptosis in cell culture. Downregulates the type 1 interferon receptor by inducing serine phosphorylation within the IFN alpha-receptor subunit 1 (IFNAR1) degradation motif and increasing IFNAR1 ubiquitination.",
        image_url: "https://www.ebi.ac.uk/pdbe/static/entry/EMD-22829/400_22829.gif",
        query_url: "/?queryId=emd-22829"
      },
    ]
  end

  def self.proteins_data
    load_data
  end

  private

  def self.get_pdb_redo_links(title, style, pdb_key, hash, keys)
    entries0 = hash && keys ? (hash.dig(*keys) || []) : []
    entries = entries0.is_a?(Hash) ? [entries0] : entries0

    with_indexes(entries, title).map do |entry, title_indexed|
      name, = entry.values_at("name")
      external_url = "https://pdb-redo.eu/db/#{name}"
      {
        title: title_indexed,
        name: name,
        style: style,
        # query_url: "/?queryId=PDB-REDO-#{pdb_key}&viewer_type=ngl&button=#query",
        query_url: "/pdb_redo/#{pdb_key}",
        external_url: external_url,
        check: external_url,
      }
    end
  end

  def self.with_indexes(entries, title)
    append_index = entries.size > 1
    entries.map.with_index(1) { |entry, idx| [entry, append_index ? "#{title} (#{idx})" : title] }
  end

  def self.get_isolde_links(title, style, pdb_key, hash, keys)
    return [] unless hash && keys
    entries = hash.dig(*keys) || []

    with_indexes(entries, title).map do |entry, title_indexed|
      entry_name, uuid = entry.values_at("name", "uuid")
      {
        title: title_indexed,
        name: entry_name,
        style: style,
        query_url: "/isolde/#{pdb_key}/#{entry_name}",
      }
    end
  end

  def self.get_refmac_links(title, style, pdb_key, hash, keys)
    return [] unless hash && keys
    entries = hash.dig(*keys) || []

    with_indexes(entries, title).map do |entry, title_indexed|
      entry_name, uuid = entry.values_at("name", "uuid")
      {
        title: title_indexed,
        name: entry_name,
        style: style,
        query_url: "/refmac/#{pdb_key}/#{entry_name}",
      }
    end
  end

  def self.get_related_keys(data)
    case
    when data.is_a?(Array)
      data
    when data.is_a?(Hash)
      data.keys
    else
      []
    end
  end

  def self.to_hash(entries)
    entries.is_a?(Hash) ? entries : entries.map { |key| [key, {}] }.to_h.sort
  end

  def self.parse_pdb(protein, keys)
    entries = protein.dig(*keys) || []
    items = []
    pockets = {}
    to_hash(entries).each do |pdb_key, pdb_hash|
      items.push({
        name: pdb_key,
        description: pdb_hash["description"],
        query_url: "/?queryId=#{pdb_key}&viewer_type=ngl&button=#query",
        image_url: "https://www.ebi.ac.uk/pdbe/static/entry/#{pdb_key}_deposited_chain_front_image-200x200.png",
        external: {text: "EBI", url: "https://www.ebi.ac.uk/pdbe/entry/pdb/#{pdb_key}"},
        api: "https://www.ebi.ac.uk/pdbe/api/pdb/entry/summary/#{pdb_key}",
        type: 'pdb',
        related: get_related_keys(pdb_hash["emdbs"]),
        pockets: pdb_hash.key?("pockets") ? pdb_hash["pockets"].map { |pocket| pocket['id']} : [],
        experiment: keys.last,
        links: [
          *get_pdb_redo_links("PDB-Redo", :turq, pdb_key, pdb_hash, ["validation", "pdb-redo"]),
          *get_isolde_links("Isolde", :cyan, pdb_key, pdb_hash, ["validation", "isolde"]),
          *get_refmac_links("Refmac", :blue_gray, pdb_key, pdb_hash, ["validation", "refmac"]),
        ],
      })

      if pdb_hash.key?("pockets")
        pdb_hash["pockets"].each do |pocket|
          id = pocket['id']
          bindingSiteScore = pocket['bindingSiteScore']
          pockets[id] = bindingSiteScore
        end
      end
    end
    return items.compact, pockets
  end

  def self.parse_emdb(protein, keys)
    entries = protein.dig(*keys) || []
    to_hash(entries).map do |emdb_key, emdb_hash|
      code = emdb_key.split("-")[1]
      {
        name: emdb_key,
        description: emdb_hash["description"],
        query_url: "/?queryId=#{emdb_key}&viewer_type=ngl&button=#query",
        image_url: "https://www.ebi.ac.uk/pdbe/static/entry/#{emdb_key}/400_#{code}.gif",
        related: get_related_keys(emdb_hash["pdbs"] || emdb_hash["pdb"]),
        external: {text: "EBI", url: "https://www.ebi.ac.uk/pdbe/entry/emdb/#{emdb_key}"},
        api: "https://www.ebi.ac.uk/pdbe/api/emdb/entry/summary/#{emdb_key}",
        type: 'emdb',
        links: [],
      }
    end
  end

  def self.parse_swiss_model(protein, keys)
    entries = protein.dig(*keys) || []
    entries.map do |entry|
      project, model, description = entry.values_at("project", "model", "description")
      {
        name: "#{project}-#{model}",
        description: ["project: #{project}", "model: #{model}", description].compact.join(" | "),
        # query_url: "/?queryId=SWISSMODEL-#{protein['uniprotAccession'][0]}-#{project}-#{model}&viewer_type=ngl&button=#query",
        query_url: "/models/#{protein['uniprotAccession'][0]}/#{keys[1]}/#{project}-#{model}",
        image_url: "https://swissmodel.expasy.org/interactive/#{project}/models/#{model}.png",
        external: {text: "SWISS-MODEL", url: "https://swissmodel.expasy.org/interactive/#{project}/models/#{model}"},
        links: [],
      }
    end
  end

  def self.parse_alphafold(protein, keys)
    entries = protein.dig(*keys) || []
    entries.map do |entry|
      name, uuid, description = entry.values_at("name", "uuid", "description")
      {
        name: name,
        description: description,
        query_url: "/models/#{protein['uniprotAccession'][0]}/#{keys[1]}/#{name}",
        image_url: "/AlphaFold_logo.png",
        external: {text: "AlphaFold", url: "https://deepmind.com/research/open-source/computational-predictions-of-protein-structures-associated-with-COVID-19"},
        links: [],
      }
    end
  end

  def self.parse_bsm_arc(protein, keys)
    entries = protein.dig(*keys) || []
    entries.map do |entry|
      model, = entry.values_at("model")
      {
        name: "#{model}",
        query_url: "/models/#{protein['uniprotAccession'][0]}/#{keys[1]}/#{model}",
        external: {text: "BSM-Arc", url: "https://bsma.pdbj.org/entry/15"},
        links: [],
      }
    end
  end

  def self.parse_db_with_experiments(protein, keys)
    entries = protein.dig(*keys) || []

    experiments = []
    items = []
    pockets = {}

    entries.each do |key,value|
      if key == 'EMDB'
        items = items + parse_emdb(protein, [*keys, "EMDB"])
      else
        new_items, pdb_pockets = parse_pdb(protein, [*keys, key])
        if new_items.size > 0 
          items = items + new_items
          experiments.push(key)
          pockets.merge!(pdb_pockets)
        end
      end
    end
    return items, experiments, pockets
    
  end

  def self.parse_db(protein, keys)
    parse_pdb(protein, [*keys, "PDB"]) + parse_emdb(protein, [*keys, "EMDB"])
  end

  def self.load_data
    json_path = File.join(__dir__, "../data/cv-data.json")
    data = JSON.parse(open(json_path).read)
    get_proteins_data(data["SARS-CoV-2 Proteome"])
  end

  def self.card(name, items, experiments = [], pockets = [])
    items.present? ? {name: name, items: items, experiments: experiments, pockets: pockets, subsections: []} : nil
  end

  def self.card_wrapper(name, subsections)
    subsections2 = subsections.compact.map { |subsection| subsection.merge(parent: name) }
    subsections2.size > 0 ? {name: name, items: [], experiments: [], pockets: [], subsections: subsections2} : nil
  end

  def self.get_relations(proteins)
    items = {}
    relations_base = proteins.each do |protein|
      protein[:sections].each do |section|
        section[:items].each do |item|
          key = item[:name] + "-" + protein[:name] + "-" + section[:name].parameterize
          items[key] = {protein: protein[:name], experiment: item[:experiment], pockets: item[:pockets]}
        end
        subsection_items = section[:subsections].flat_map do |subsection|
          subsection[:items].each do |item|
            key = item[:name] + "-" + protein[:name] + "-" + subsection[:name].parameterize
            items[key] = {protein: protein[:name], experiment: item[:experiment], pockets: item[:pockets]}
          end
        end
      end
    end

    return items
  end

  def self.get_proteins_data(proteins_raw)
    proteins = proteins_raw.map do |name, protein|
      {
        name: name,
        names: protein["names"],
        description: protein["description"],
        polyproteins: protein["uniprotAccession"],
        sections: [
          card("PDB", *parse_pdb(protein, ["PDB"])),
          card("EMDB", parse_emdb(protein, ["EMDB"])),
          card_wrapper("Interactions", [
            card("Protein-Protein Interactions", *parse_db_with_experiments(protein, ["Interactions", "P-P-Interactions"])),
            card("Ligands", *parse_db_with_experiments(protein, ["Interactions", "Ligands"])),
          ]),
          card_wrapper("Related", [
            card("SARS-CoV", *parse_db_with_experiments(protein, ["Related", "SARS-CoV"])),
            card("Other", *parse_db_with_experiments(protein, ["Related", "OtherRelated"]))
          ]),
          card_wrapper("Computational Models", [
            card("Swiss Model", parse_swiss_model(protein, ["CompModels", "swiss-model"])),
            card("AlphaFold", parse_alphafold(protein, ["CompModels", "AlphaFold"])),
            card("BSM-Arc", parse_bsm_arc(protein, ["CompModels", "BSM-Arc"])),
          ]),

        ].compact,
      }
    end

    {proteins: proteins, relations: get_relations(proteins)}
  end
end
